planner/core/find_best_task.go

```diff
@@ -407,17 +420,163 @@ func (ds *DataSource) findBestTask(prop *property.PhysicalProperty) (t task, err
+				isPartition:     ds.isPartition,
+				physicalTableID: ds.physicalTableID,
+			}.Init(ds.ctx)
+			ts.SetSchema(ds.schema)
+			if ts.Table.PKIsHandle {
+				if pkColInfo := ts.Table.GetPkColInfo(); pkColInfo != nil {
+					if ds.statisticTable.Columns[pkColInfo.ID] != nil {
+						ts.Hist = &ds.statisticTable.Columns[pkColInfo.ID].Histogram
+					}
+				}
+			}
+			path := candidate.path
+			ts.Ranges = path.ranges
+			ts.AccessCondition, ts.filterCondition = path.accessConds, path.tableFilters
+			rowCount := path.countAfterAccess
+			ts.stats = property.NewSimpleStats(rowCount)
+			ts.stats.StatsVersion = ds.statisticTable.Version
+			if ds.statisticTable.Pseudo {
+				ts.stats.StatsVersion = statistics.PseudoVersion
+			}
+			totalRowCount += rowCount
+			partPaths = append(partPaths, ts)
+			continue
+		}
+		idx := partPath.index
+		is := PhysicalIndexScan{
+			Table:            ds.tableInfo,
+			TableAsName:      ds.TableAsName,
+			DBName:           ds.DBName,
+			Columns:          ds.Columns,
+			Index:            idx,
+			IdxCols:          path.idxCols,
+			IdxColLens:       path.idxColLens,
+			AccessCondition:  path.accessConds,
+			Ranges:           path.ranges,
+			dataSourceSchema: ds.schema,
+			isPartition:      ds.isPartition,
+			physicalTableID:  ds.physicalTableID,
+		}.Init(ds.ctx)
+		statsTbl := ds.statisticTable
+		if statsTbl.Indices[idx.ID] != nil {
+			is.Hist = &statsTbl.Indices[idx.ID].Histogram
+		}
+		is.initSchema(ds.id, idx, true)
+		rowCount := partPath.countAfterAccess
+		if prop.IsEmpty() && prop.ExpectedCnt < ds.stats.RowCount {
+			selectivity := ds.stats.RowCount / path.countAfterAccess
+			rowCount = math.Min(prop.ExpectedCnt/selectivity, rowCount)
+		}
+		is.stats = property.NewSimpleStats(rowCount)
+		is.stats.StatsVersion = ds.statisticTable.Version
+		if ds.statisticTable.Pseudo {
+			is.stats.StatsVersion = statistics.PseudoVersion
+		}
+		totalRowCount += rowCount
+		partPaths = append(partPaths, is)
+	}
+	ts := PhysicalTableScan{
+		Table:           ds.tableInfo,
+		Columns:         ds.Columns,
+		TableAsName:     ds.TableAsName,
+		DBName:          ds.DBName,
+		isPartition:     ds.isPartition,
+		physicalTableID: ds.physicalTableID,
+	}.Init(ds.ctx)
+	ts.SetSchema(ds.schema)
+	if ts.Table.PKIsHandle {
+		if pkColInfo := ts.Table.GetPkColInfo(); pkColInfo != nil {
+			if ds.statisticTable.Columns[pkColInfo.ID] != nil {
+				ts.Hist = &ds.statisticTable.Columns[pkColInfo.ID].Histogram
+			}
+		}
+	}
+	partPathsConds := make([]expression.Expression, 0, 1)
+	for _, cond1 := range ds.pushedDownConds {
+		find := true
+		for _, cond2 := range path.tableFilters {
+			if cond1 == cond2 {
+				find = false
+				break
+			}
+		}
+		if find {
+			partPathsConds = append(partPathsConds, cond1)
+		}
+	}
+	selectivity, _, err := ds.stats.HistColl.Selectivity(ds.ctx, partPathsConds)
+	if err != nil {
+		logutil.BgLogger().Debug("calculate selectivity failed, use selection factor", zap.Error(err))
```

Comment:
Are you forget to delete the debug information?

Reply:
By the way, can you add some related tests?

Reply:
> Are you forget to delete the debug information?

It is not. This tells if failed using Selectivity(). I think it is needed.

Reply:
> By the way, can you add some related tests?
This version is used to confirm whether the implement  is ok and what shortages it has. So, for the moment, it lacks test.